Promised file data should contain UTType identifier instead of file extension

aprotyas opened the pull request:
#### 2db314d62ff04fd0058c80d219cfc3f8b6374942
<pre>
Promised file data should contain UTType identifier instead of file extension
<a href="https://bugs.webkit.org/show_bug.cgi?id=254555">https://bugs.webkit.org/show_bug.cgi?id=254555</a>
rdar://104449116

Reviewed by Wenson Hsieh.

When dragging an image in Safari, we notice that the file promise
provider specifies the file type as _just_ the extension string, when
it should be specifying the Universal Type Identifier (UTI) instead,
e.g. `public.png` instead of `png`.

This patch corrects said mismatch by supplying the UTI instead of the
file extension as the representation for `NSFilesPromisePboardType`.

* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::setPromisedDataForImage):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setFileAndURLTypes):
(WebKit::WebViewImpl::setPromisedDataForImage):
* Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/262280@main">https://commits.webkit.org/262280@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73b408dbf7b37d76bc77e7905d7a4d5f0606f5f4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1064 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1093 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1674 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/940 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1048 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1142 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1173 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/1135 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1071 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/995 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/989 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1557 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/1036 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/987 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/975 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/970 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1015 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/2050 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1015 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/949 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/957 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/996 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/270 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1032 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4198,7 +4198,7 @@ static bool matchesExtensionOrEquivalent(const String& filename, const String& e
     [types addObjectsFromArray:archiveBuffer ? PasteboardTypes::forImagesWithArchive() : PasteboardTypes::forImages()];
     [pasteboard declareTypes:types.get() owner:m_view.getAutoreleased()];
-    setFileAndURLTypes(filename, extension, title, url, visibleURL, pasteboard);
+    setFileAndURLTypes(filename, extension, image->uti(), title, url, visibleURL, pasteboard);
```
is `image` guaranteed to be non-null here?

whsieh replied:
I'm pretty sure it is, since the only call site:

```
void PageClientImpl::setPromisedDataForImage(const String& pasteboardName, Ref<FragmentedSharedBuffer>&& imageBuffer, const String& filename, const String& extension, const String& title, const String& url, const String& visibleURL, RefPtr<FragmentedSharedBuffer>&& archiveBuffer, const String& originIdentifier)
{
    auto image = BitmapImage::create();
    image->setData(WTFMove(imageBuffer), true);
    m_impl->setPromisedDataForImage(image.ptr(), filename, extension, title, url, visibleURL, archiveBuffer.get(), pasteboardName, originIdentifier);
}
```

...passes in an `Image*` from a `Ref<BitmapImage>`. A couple of observations on that note:

- The `Image*` should probably be changed to an `Image&` to reflect this.
- There's code above that already grabs the `Image::uti()`; maybe we can store that into a local variable and use it in both places.

aprotyas replied:
Yes, `image` is guaranteed to be non-null here as @whsieh posits. I've also refactored a little bit (https://github.com/WebKit/WebKit/commit/73b408dbf7b37d76bc77e7905d7a4d5f0606f5f4) to:

- Pass in a `WebCore::Image&` to `WebViewImpl::setPromisedDataForImage`
- Store `WebCore::Image::uti()` in a local variable.

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
```diff
@@ -114,6 +115,13 @@
+    TestDraggingInfo *draggingInfo = [simulator draggingInfo];
+    NSArray<NSFilePromiseReceiver *> *filePromiseReceivers = [draggingInfo filePromiseReceivers];
+    EXPECT_EQ(1UL, [filePromiseReceivers count]);
+    NSFilePromiseReceiver *filePromiseReceiver = [filePromiseReceivers objectAtIndex:0];
```
```suggestion
    NSFilePromiseReceiver *filePromiseReceiver = filePromiseReceivers.firstObject;
```

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/73b408dbf7b37d76bc77e7905d7a4d5f0606f5f4

pxlcoder commented on Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
```diff
@@ -114,6 +115,13 @@
+    EXPECT_EQ(1UL, [filePromiseReceivers count]);
+    NSFilePromiseReceiver *filePromiseReceiver = [filePromiseReceivers objectAtIndex:0];
+    EXPECT_EQ(1UL, [filePromiseReceiver.fileTypes count]);
+    EXPECT_WK_STREQ((__bridge NSString *)kUTTypeGIF, [filePromiseReceiver.fileTypes objectAtIndex:0]);
```
```suggestion
    EXPECT_WK_STREQ((__bridge NSString *)kUTTypeGIF, filePromiseReceiver.fileTypes.firstObject);
```

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/73b408dbf7b37d76bc77e7905d7a4d5f0606f5f4

pxlcoder reviewed (commented):
Two nits, since the patch is going to be updated anyway 😅

aprotyas committed:
Promised file data should contain UTType identifier instead of file extension
https://bugs.webkit.org/show_bug.cgi?id=254555
rdar://104449116

Reviewed by Wenson Hsieh.

When dragging an image in Safari, we notice that the file promise
provider specifies the file type as _just_ the extension string, when
it should be specifying the Universal Type Identifier (UTI) instead,
e.g. `public.png` instead of `png`.

This patch corrects said mismatch by supplying the UTI instead of the
file extension as the representation for `NSFilesPromisePboardType`.

* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::setPromisedDataForImage):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setFileAndURLTypes):
(WebKit::WebViewImpl::setPromisedDataForImage):
* Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
(TEST):

Canonical link: https://commits.webkit.org/262280@main
